nexus/src/db/datastore/silo_group.rs

```diff
@@ -0,0 +1,208 @@
+    /// - remove the user from groups if they no longer have membership
+    ///
+    /// Do this as one transaction that deletes all current memberships for a
+    /// user, then adds back the ones they are in. This avoids the scenario
+    /// where a crash half way through causes the resulting group memberships to
+    /// be incorrect.
+    pub async fn silo_group_membership_replace_for_user(
+        &self,
+        opctx: &OpContext,
+        silo_user_id: Uuid,
+        silo_group_ids: Vec<Uuid>,
+    ) -> UpdateResult<()> {
+        let (_authz_silo_user, ..) = LookupPath::new(opctx, &self)
+            .silo_user_id(silo_user_id)
+            .fetch_for(authz::Action::Modify)
+            .await?;
+
+        self.pool_authorized(opctx)
+            .await?
+            .transaction(move |conn| {
+                use db::schema::silo_group_membership::dsl;
+
+                // Delete existing memberships for user
+                diesel::delete(dsl::silo_group_membership)
+                    .filter(dsl::silo_user_id.eq(silo_user_id))
+                    .execute(conn)?;
+
+                // Create new memberships for user
+                let silo_group_memberships: Vec<
+                    db::model::SiloGroupMembership,
+                > = silo_group_ids
+                    .iter()
+                    .map(|group_id| db::model::SiloGroupMembership {
+                        silo_group_id: *group_id,
+                        silo_user_id: silo_user_id,
+                    })
+                    .collect();
+
+                diesel::insert_into(dsl::silo_group_membership)
+                    .values(silo_group_memberships)
+                    .execute(conn)?;
+
+                Ok(())
+            })
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+    }
+
+    pub async fn silo_group_delete(
+        &self,
+        opctx: &OpContext,
+        authz_silo_group: &authz::SiloGroup,
+    ) -> DeleteResult {
+        opctx.authorize(authz::Action::Delete, authz_silo_group).await?;
+
+        #[derive(Debug, thiserror::Error)]
+        enum SiloDeleteError {
+            #[error("group {0} still has memberships")]
+            GroupStillHasMemberships(Uuid),
+        }
+        type TxnError = TransactionError<SiloDeleteError>;
+
+        let group_id = authz_silo_group.id();
+
+        self.pool_authorized(opctx)
+            .await?
+            .transaction(move |conn| {
+                use db::schema::silo_group_membership;
+
+                // Don't delete groups that still have memberships
+                let group_memberships =
+                    silo_group_membership::dsl::silo_group_membership
+                        .filter(
+                            silo_group_membership::dsl::silo_group_id
+                                .eq(group_id),
+                        )
+                        .select(SiloGroupMembership::as_returning())
```

Comment:
I'd add a `limit 1` here in case the group is huge.

Reply:
Good idea, we're only checking if any group membership exists. Added in a2507a60.